### PR TITLE
MKLDNN 1.0.2

### PR DESCRIPTION
**Pull request by sreekanth-yalachigere**
Upgrading MKL-DNN EP to MKL-DNN 1.0.2

### CODE REVIEW

File: cmake/external/mkldnn.cmake [hunk 1]
```diff
@@ -1,41 +1,28 @@
 include (ExternalProject)
 
 set(MKLDNN_URL https://github.com/intel/mkl-dnn.git)
-# If MKLDNN_TAG is updated, check if MKLML_VERSION and platform.cmake.patch need to be updated.
-set(MKLDNN_TAG v0.18.1)
-set(MKLML_VERSION 2019.0.5.20190502)
+set(MKLDNN_TAG v1.0.2)
```
Comment (jywu-msft) [inline cmake/external/mkldnn.cmake]: please don't remove the MKLML portions.
while mklml isn't used by mkldnn explicitly, there is a build option for using mklml by itself. 
and we produce a ort mkl nuget package for customers who want to use mklml only.

File: cmake/external/mkldnn.cmake [hunk 2]
```diff
@@ -1,41 +1,28 @@
...
   endif()
   if(onnxruntime_USE_MKLML)
-    set(MKLML_SHARED_LIB libmklml_intel.so)
     set(IOMP5MD_SHARED_LIB libiomp5.so)
   endif()
 endif()
 
 if (onnxruntime_USE_MKLML)
-  set(MKLDNN_VERSION_SHORT v0.20)
-  set(MKLML_URL https://github.com/intel/mkl-dnn/releases/download/${MKLDNN_VERSION_SHORT}/mklml_${MKLML_OS_VERSION_STR}_${MKLML_VERSION}.${MKLML_FILE_EXTENSION})
```
Comment (jywu-msft) [inline cmake/external/mkldnn.cmake]: don't remove this. make sure --use_mklml still works (by itself, without --use_mkldnn)

File: cmake/external/mkldnn.cmake (hunk 1 above)
Comment (sreekanth-yalachigere) [reply cmake/external/mkldnn.cmake]: added MKLML back to cmake.

File: cmake/external/mkldnn.cmake [hunk 3]
```diff
@@ -34,7 +34,7 @@ else()
 endif()
 
 if (onnxruntime_USE_MKLML)
-  set(MKLDNN_VERSION_SHORT v0.20)
+  set(MKLDNN_VERSION_SHORT v1.0)
   set(MKLML_URL https://github.com/intel/mkl-dnn/releases/download/${MKLDNN_VERSION_SHORT}/mklml_${MKLML_OS_VERSION_STR}_${MKLML_VERSION}.${MKLML_FILE_EXTENSION})
```
Comment (jywu-msft) [inline cmake/external/mkldnn.cmake]: The mklml binary isn't available with v1.0 so we need to use the previous version for the mklml url

Comment (sreekanth-yalachigere) [reply cmake/external/mkldnn.cmake]: my bad. I changed MKLDNN_VERSION_SHORT to v0.20 so that we  we can download mklml. double checked with --use_mklml

File: onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc [hunk 4]
```diff
@@ -197,7 +170,7 @@ std::vector<std::unique_ptr<ComputeCapability>> MKLDNNExecutionProvider::GetCapa
   ORT_UNUSED_PARAMETER(kernel_registries);
 
   // temporary switch to toggle between mkldnn-vanilla and mkldnn-subgraph implementation using
-  // ORT_MKLDNN_SUBGRAPH environment variable
+  //
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.cc]: get rid of this line

File: onnxruntime/core/providers/mkldnn/subgraph/mkldnn_activations.h [hunk 5]
```diff
@@ -27,7 +27,7 @@ class MklDnnRelu : public MklDnnKernel {
                           OrtKernelContext* context,
                           mkldnn::engine& cpu_engine,
                           std::vector<mkldnn::primitive>& net,
-                          mkldnn::memory::format& source_format) {
+                          std::vector<std::unordered_map<int, mkldnn::memory>> &net_args) {
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/subgraph/mkldnn_activations.h]: nit: & should be after >>

File: onnxruntime/core/providers/mkldnn/subgraph/subgraph.h [hunk 6]
```diff
@@ -48,8 +48,12 @@ struct Subgraph {
     std::vector<std::string> outputs;
     std::vector<std::string> outputs_as_input_other_node;
     std::vector<onnxruntime::NodeIndex> subgraph_node_indexes;
+    int subgraph_index = 0;
 
-	void Reset() {
+    SubgraphVariables() {
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/subgraph/subgraph.h]: why is this needed? i thought we got rid of it in https://github.com/microsoft/onnxruntime/pull/1740

**Comment by jywu-msft** [comment]
looks like the windows build fails.

**Comment by sreekanth-yalachigere** [comment]
> looks like the windows build fails.

Looking at it. cloned MKL-DNN 1.0.2 is failing. It builds for me on my dev box.

File: onnxruntime/core/providers/mkldnn/mkldnn_common.h [hunk 7]
```diff
@@ -5,10 +5,13 @@
 #include "core/common/common.h"
 #include "mkldnn.hpp"
 #include <unordered_map>
+#include <list>
 
 namespace onnxruntime {
 namespace mkl_dnn {
 
+#define MKLDNN_EP_LRU_CACHE_DEFAULT_SIZE 500
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/mkldnn_common.h]: this will be implemented later right? remove for now.

File: onnxruntime/core/providers/mkldnn/subgraph/mkldnn_activations.h [hunk 8]
```diff
@@ -44,25 +44,27 @@ class MklDnnRelu : public MklDnnKernel {
...
+	  x_shape = TensorShape(xshape, xdim);
 
-      mkldnn::memory::dims src_dims_mkl(
+      mkldnn::memory::dims src_dims(
           x_shape.GetDims().begin(), x_shape.GetDims().end());
 
+      ort_source_desc_ = mkldnn::memory::desc(
+          {src_dims}, MklDnnType<T>(), ort_source_format_);
+      source_desc_ = ort_source_desc_;
       src_md_.reset(new mkldnn::memory::desc(
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/subgraph/mkldnn_activations.h]: we can take this opportunity to clean up the smart pointer usage in all these files.
src_md_ = make_shared<T>(...) is preferred and more efficient than reset(new T)
let's clean all this up. thanks.

Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/subgraph/mkldnn_activations.h]: Most of the pointers are unique_ptr and we need make_unique.
I was about to push my changes. I will make these changes and then create a new PR

Comment (jywu-msft) [reply onnxruntime/core/providers/mkldnn/subgraph/mkldnn_activations.h]: > Most of the pointers are unique_ptr and we need make_unique.
> I was about to push my changes. I will make these changes and then create a new PR

let's do the cleanup as a separate PR then. priority is to get this one passing the CI's and merged soon.

Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/subgraph/mkldnn_activations.h]: okay. I will close this and create new new PR. As I have merged latest from master repo.

**Comment by sreekanth-yalachigere** [comment]
Creating new PR
